[TensorExpr] Add a class for representing data type.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33220 [TensorExpr] Add IR Printer.
* #33219 [TensorExpr] Add IR visitor, IR mutator, and IR evaluator.
* #33218 [TensorExpr] Add core classes for representing expressions and statements.
* **#33217 [TensorExpr] Add a class for representing data type.**
* #33216 [TensorExpr] Add classes for memory management in tensor expressions.



Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

Commit:
[TensorExpr] Add a class for representing data type.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

ZolotukhinM commented:
This commit also adds tests - they are currently implemented exactly like `test_jit` tests.

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

dzhulgakov commented on test/cpp/tensorexpr/README.md:
````diff
@@ -0,0 +1,55 @@
+Then, register your test in `tests.h`:
+```cpp
+// Add to TH_FORALL_TESTS_CUDA instead for CUDA-requiring tests
+#define TH_FORALL_TESTS(_)             \
````
why do we do this instead of just using gtest? (out of curiosity)

is it only because of CUDA?

ZolotukhinM replied:
I copied the test infra from JIT tests almost verbatim.

dzhulgakov replied:
ok, there might be a better setup, but we can follow the JIT practice directly for now

dzhulgakov commented on test/cpp/tensorexpr/tests_setup.py:
```diff
@@ -0,0 +1,88 @@
+import sys
```
what is this magic file?

ZolotukhinM replied:
It was also copied, we probably don't need it :)

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Looks good. Also it's been reviewed on another branch.

Commit:
Rebase on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add a class for representing data type."


Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add a class for representing data type."




Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]

Commit:
Rebase. on "[TensorExpr] Add a class for representing data type."




Differential Revision: [D19848380](https://our.internmc.facebook.com/intern/diff/D19848380)

[ghstack-poisoned]
